nvflare/fuel/f3/cellnet/cell.py

```diff
@@ -0,0 +1,1757 @@
+        }
+
+    @staticmethod
+    def from_dict(d: dict):
+        msg_dict = d.get("message")
+        msg = new_message(headers=msg_dict.get("headers"), payload=msg_dict.get("payload"))
+        return TargetMessage(target=d.get("target"), channel=d.get("channel"), topic=d.get("topic"), message=msg)
+
+
+class CellAgent:
+    """
+    A CellAgent represents a cell in another cell.
+    """
+
+    def __init__(self, fqcn: str, endpoint: Endpoint):
+        """
+
+        Args:
+            fqcn: FQCN of the cell represented
+        """
+        err = FQCN.validate(fqcn)
+        if err:
+            raise ValueError(f"Invalid FQCN '{fqcn}': {err}")
+
+        self.info = FqcnInfo(FQCN.normalize(fqcn))
+        self.endpoint = endpoint
+
+    def get_fqcn(self):
+        return self.info.fqcn
+
+
+class _CB:
+    def __init__(self, cb, args, kwargs):
+        self.cb = cb
+        self.args = args
+        self.kwargs = kwargs
+
+
+class _Registry:
+    def __init__(self):
+        self.reg = {}  # channel/topic => _CB
+
+    @staticmethod
+    def _item_key(channel: str, topic: str) -> str:
+        return f"{channel}:{topic}"
+
+    def set(self, channel: str, topic: str, item):
+        key = self._item_key(channel, topic)
+        self.reg[key] = item
+
+    def append(self, channel: str, topic: str, item):
+        key = self._item_key(channel, topic)
+        item_list = self.reg.get(key)
+        if not item_list:
+            item_list = []
+            self.reg[key] = item_list
+        item_list.append(item)
+
+    def find(self, channel: str, topic: str):
+        item = self.reg.get(self._item_key(channel, topic))
+        if not item:
+            # try topic * in channel
+            item = self.reg.get(self._item_key(channel, "*"))
+
+        if not item:
+            # try topic * in channel *
+            item = self.reg.get(self._item_key("*", "*"))
+
+        return item
+
+
+class _Waiter(threading.Event):
+    def __init__(self, targets: List[str]):
+        super().__init__()
+        self.targets = [x for x in targets]
+        self.reply_time = {}  # target_id => reply recv timestamp
+        self.send_time = time.time()
+        self.id = str(uuid.uuid4())
+        self.received_replies = {}
+
+
+class _BulkSender:
+    def __init__(self, cell, target: str, max_queue_size):
+        self.cell = cell
+        self.target = target
+        self.max_queue_size = max_queue_size
+        self.messages = []
+        self.last_send_time = 0
+        self.lock = threading.Lock()
+
+    def queue_message(self, channel: str, topic: str, message: Message):
+        encode_payload(message)
+        with self.lock:
+            tm = TargetMessage(target=self.target, channel=channel, topic=topic, message=message)
+            self.messages.append(tm)
+            self.cell.logger.debug(f"{self.cell.get_fqcn()}: bulk sender {self.target} queue size {len(self.messages)}")
```

Comment:
This uses the self.cell.logger, why not create its own logger?

Reply:
Added a logger for the class.